Complete work on etcd sync

miracle2k opened the pull request:
Based on my previous comment (https://github.com/flynn/strowger/pull/4#issuecomment-37291917), here is an attempt to implement the sync with etcd.

It's 90% complete, but since I've had to make a growing number of design choices, I thought it best to check   with you guys.

Crucially, what I am doing here, somewhat independently from the sync feature, is making the external API a simple wrapper around etcd - it does not touch the internal data structures. Those are updated only via the update stream from etcd. 

This is so to not run the internal-apply code twice (I only realize now that it might have been possible to detect an update originating from the self instance via isLeader() or by looking at the etcd indices).

Relying on the update stream has a certain simple beauty to it, it makes sure the sync-code has to work properly even when a single strowger instance is used, which seems quite positive, though it raises a question for me:

Errors like "domain already exists" - can we implement error handling like that on top of the etcd interface, or should the API provide  simple "set and overwrite only" functionality?

miracle2k committed:
Complete work on etcd sync.

Signed-off-by: Michael Elsdörfer <michael@elsdoerfer.com>

titanous commented:
Would you rebase so that only the sync commit shows up?

miracle2k commented:
Done.

titanous commented on http.go:
```diff
@@ -60,20 +61,73 @@ func NewHTTPFrontend(addr, tlsAddr string, etcdc EtcdClient, discoverdc Discover
 func (s *HTTPFrontend) AddHTTPDomain(domain string, service string, cert []byte, key []byte) error {
-	return s.addDomain(domain, service, cert, key, true)
+	// XXX: Is the API working with etcd directly good enough? How about validating things?
+	if _, err := s.etcd.Create(s.etcdPrefix+domain+"/service", service, 0); err != nil {
```
I think this should use `etcd.Create` and return `ErrDomainExists` if it fails due to the key existing.

titanous commented on http.go:
```diff
@@ -60,20 +61,73 @@ func NewHTTPFrontend(addr, tlsAddr string, etcdc EtcdClient, discoverdc Discover
+	if _, err := s.etcd.Create(s.etcdPrefix+domain+"/tls/key", string(key), 0); err != nil {
+		return err
+	}
+	if _, err := s.etcd.Create(s.etcdPrefix+domain+"/tls/cert", string(cert), 0); err != nil {
```
Can we only do this when the key/cert are set?

titanous commented:
Yeah, I like this approach. Should I wait until it's 100% to review?

miracle2k commented:
Yes, unless this isn't going in the wrong direction entirely, I'll finish it first.

titanous commented:
:+1:

miracle2k committed:
Create tls/key and tls/cert only when used.

miracle2k committed:
Make HTTP serving code aware of incomplete frontends.

miracle2k commented:
I think that is it from my part.

titanous commented on http.go:
```diff
@@ -60,20 +62,87 @@ func NewHTTPFrontend(addr, tlsAddr string, etcdc EtcdClient, discoverdc Discover
 func (s *HTTPFrontend) AddHTTPDomain(domain string, service string, cert []byte, key []byte) error {
-	return s.addDomain(domain, service, cert, key, true)
+	var err error
+	var _ etcd.Response
```
This line is a no-op.

titanous commented on http.go:
```diff
@@ -60,20 +62,87 @@ func NewHTTPFrontend(addr, tlsAddr string, etcdc EtcdClient, discoverdc Discover
+			goto error
+		}
+	}
+	if len(cert) > 0 {
```
It should be an error for one of these to be empty, so wrap both creates in the same if statement? (`len(cert) > 0 && len(key) > 0`)

titanous commented on http.go:
```diff
@@ -60,20 +62,87 @@ func NewHTTPFrontend(addr, tlsAddr string, etcdc EtcdClient, discoverdc Discover
+	}
+	return nil
+error:
+	if err.(*etcd.EtcdError).ErrorCode == 105 {
```
``` go
if e, ok := err.(*etcd.EtcdError); ok && e.ErrorCode == 105 {
```

titanous commented on http.go:
```diff
@@ -60,20 +62,87 @@ func NewHTTPFrontend(addr, tlsAddr string, etcdc EtcdClient, discoverdc Discover
+
+func (s *HTTPFrontend) RemoveHTTPDomain(domain string) error {
+	if _, err := s.etcd.Delete(s.etcdPrefix+domain, true); err != nil {
+		if err.(*etcd.EtcdError).ErrorCode == 100 {
```
Ditto on the type check.

titanous commented on http.go:
```diff
@@ -60,20 +62,87 @@ func NewHTTPFrontend(addr, tlsAddr string, etcdc EtcdClient, discoverdc Discover
+	s.mtx.Lock()
+	d, ok := s.domains[name]
+	if !ok {
+		return ErrNoSuchDomain
```
Needs a `s.mtx.Unlock()`

titanous replied:
Easier would be to just put the unlock above the `if !ok {`.

miracle2k replied:
The reason defer isn't used here is the call to setDomainService() below, which it turn locks itself, and the lock isn't re-entrant. The reason setDomainService() is called during domain removal is to trigger the service refcount cleanup code, which I wanted to share in this way.

titanous replied:
Yep makes sense, just make sure that you unlock before returning in the error case.

titanous commented on http.go:
```diff
@@ -60,20 +62,87 @@ func NewHTTPFrontend(addr, tlsAddr string, etcdc EtcdClient, discoverdc Discover
+	}
+	s.mtx.Unlock()
+
+	err := s.setDomainService(d, "")
```
I don't see a special case in `setDomainService` for an empty service name, what happens in this case?

titanous commented on http.go:
```diff
@@ -60,20 +62,87 @@ func NewHTTPFrontend(addr, tlsAddr string, etcdc EtcdClient, discoverdc Discover
+	if d.server != nil {
+		d.server.refs--
+		if d.server.refs <= 0 {
+			// TODO: close service set stream
```
Resolve this TODO?

titanous commented on http.go:
```diff
@@ -85,51 +154,46 @@ func (s *HTTPFrontend) addDomain(name string, service string, cert []byte, key [
+	if len(d.cert) > 0 && len(d.key) > 0 {
+		var keypair *tls.Certificate
+		if cert != nil && key != nil {
+			created, err := tls.X509KeyPair(cert, key)
```
Could do this:

``` go
var err error
d.keypair, err = tls.X509KeyPair(cert, key)
```

miracle2k replied:
I don't think this works because domain.keypair is a pointer.

titanous replied:
Ah, my bad.

titanous commented on http.go:
```diff
@@ -168,7 +232,13 @@ func (s *HTTPFrontend) syncDatabase() {
+		}
+
 		// Recursively get the whole service structure
+		// XXX Get the whole structure recursively up front, or fix to since index?
```
Should this be resolved? I can help if I have more context.

miracle2k replied:
I believe that is is, if changes are made to etcd after the initial Get() to fetch the domain list, but before Watch() is called with the index from that initial interaction, those changes can be applied twice: Once during the subsequent Get() calls to fetch the data for each domain, then as a notification through Watch().

The two solutions I could think of were using a recursive call to fetch the whole structure at once, or use the index given to us by the root Get() for all subsequent data fetches as well.

titanous replied:
I prefer the recursive fetch, but if you think using the index would be cleaner that's fine too.

miracle2k replied:
I was only thinking performance for large datasets, you'd be streaming the sync process essentially, but I really have no idea. Recursive is certainly more straightforward to do.

titanous replied:
etcd has no pagination, so that feels like premature optimization.

miracle2k committed:
Minor improvements.

miracle2k committed:
Properly allow removing the service configuration.

miracle2k committed:
Fix missing unlock.

This is ugly, but I don't see an alternative except using a separate lock for the s.services field.

miracle2k committed:
Remove TODO.

miracle2k committed:
Close the service set connection on remove.

miracle2k committed:
Get initial etcd config in one atomic operation.

titanous commented:
@miracle2k What do you think about serializing the config as JSON and saving it in a single key per domain? (as described in https://github.com/flynn/strowger/issues/11#issuecomment-37822044)

I think doing that would also make it easy to extract out a separate package that does config sync and reuse it for the TCP implementation (and in the future UDP).

titanous commented:
@miracle2k I'm going to merge this and do some more improvements. Would you rebase/squash?

miracle2k commented:
Apologies. I'm in Internet no-mans land right now, and I do not have my regular setup either. I could probably figure out the squash, but it would be significantly easier if you could do it.

titanous commented:
@miracle2k sure, I'll take care of it.
